paddle/fluid/framework/details/inplace_op_pass.cc

```diff
@@ -83,490 +78,374 @@ const std::string kInplacedOpWhiteList[] = { // NOLINT
+                 out_var_name);
+
+  // The following codes are designed to guarantee that ssa_map_ is still
+  // an ssa graph after inplace is performed.
+  // Step 1: Rename the following versions of out_var as the name of in_var
+  // Step 2: Remove the following versions of out_var and append them to in_var
+  // Be careful that the inputs of input op of out_var should not be renamed,
+  // but outputs should be renamed.
+  auto original_iter = iter;
+  while (iter != all_out_nodes.end()) {
+    auto *node = *iter;
+    /* Step 1 */
+    node->RenameVar(in_var_name);
+    if (iter != original_iter) {
+      for (auto *in : node->inputs) {
+        if (in->IsOp() && in->Op()) {
+          in->Op()->RenameOutput(out_var_name, in_var_name);
+          in->Op()->RenameInput(out_var_name, in_var_name);
+          in->Op()->Flush();
         }
-
-        swap_nodes.emplace_back(std::make_pair(node, cache_node));
       }
     }
 
-    // if we need to rename the output,
-    // always create a newer version of cache_var
-    for (auto* node : op->outputs) {
-      if (node->Name() == var) {
-        ir::Node* cache_node = graph->CreateVarNode(var_desc.get());
-        // swap node to cache node
-        cache_node->outputs.insert(cache_node->outputs.end(),
-                                   node->outputs.begin(), node->outputs.end());
-        cache_node->inputs.emplace_back(op);
-        std::replace(op->outputs.begin(), op->outputs.end(), node, cache_node);
-        for (auto* next_op : node->outputs) {
-          std::replace(next_op->inputs.begin(), next_op->inputs.end(), node,
-                       cache_node);
-        }
-
-        swap_nodes.emplace_back(std::make_pair(node, cache_node));
+    for (auto *out : node->outputs) {
+      if (out->IsOp() && out->Op()) {
+        out->Op()->RenameOutput(out_var_name, in_var_name);
+        out->Op()->RenameInput(out_var_name, in_var_name);
+        out->Op()->Flush();
       }
     }
+
+    /* Step 2 */
+    all_in_nodes.emplace_back(node);
+    ++iter;
   }
 
-  return swap_nodes;
-}
+  /* Step 2 */
+  all_out_nodes.erase(original_iter, all_out_nodes.end());
 
-void InplacePass::CommitModify(const NodeSwapQueue& swap_nodes,
-                               ir::Graph* graph) const {
-  for (auto& pair : swap_nodes) {
-    auto *node = pair.first, *cache_node = pair.second;
-    const std::string var = node->Name(), cache_var = cache_node->Name();
-    var_nodes_[cache_var].emplace_back(cache_node);
-    graph->RemoveNode(node);
-    auto& nodes = var_nodes_.at(var);
-    // release unused var in graph. Because python side memory optimize
-    // may reused the var in same name, so we only clear the var node
-    // after current inplaced index.
-    nodes.erase(std::remove(nodes.begin(), nodes.end(), node), nodes.end());
+  if (all_out_nodes.empty()) {
+    ssa_map_.erase(out_var_name);
   }
+  op->Op()->RenameOutput(out_var_name, in_var_name);
```

Comment:
I saw you did Op()->Flush() after renaming at this function above. Do we need to do Flush() after renaming here?

Reply:
Done.